### DISCUSSION TITLE
Why is`PFN_vkVoidFunction` in a `Option`?

### DISCUSSION THREAD

**Discussion by Local-Trash** [KyleMayes/vulkanalia#239] [Q&A] [answered]
**Note:** This is my first post in a discussion. Sorry if I mess up or anything. Please tell if I did something wrong.

I have been using the `vulkanalia-sys` crate and I have also tried out the `ash` crate. Through this time, I have notice that some of the function pointers are wrapped in a `Option` and that they are not pointers themselves(e.i. type x = *mut y). I was wondering why this is in the `sys` crate. Thank you for your time.

**Comment by KyleMayes** [comment, marked as answer]
The Rust type:

https://github.com/KyleMayes/vulkanalia/blob/384a8d39468e347c24c9b862e5568f06ac0b469e/vulkanalia-sys/src/functions.rs#L87

The C type (according to the [Vulkan specification](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/PFN_vkVoidFunction.html#_c_specification)):

```c
typedef void (VKAPI_PTR *PFN_vkVoidFunction)(void);
```

The reason the Rust type is a `Option<unsafe extern "system" fn()>` instead of just `unsafe extern "system" fn()` is to allow it to represent null function pointers. C function pointer types implicitly allow null values, but Rust function pointer types do not.

The Rust types for actual Vulkan commands in `vulkanalia` do not include `Option`, e.g.:

https://github.com/KyleMayes/vulkanalia/blob/384a8d39468e347c24c9b862e5568f06ac0b469e/vulkanalia-sys/src/commands.rs#L3041-L3042
